### DISCUSSION TITLE
Error on Services registration -> require(...).default is not a constructor

### DISCUSSION THREAD

**Discussion by lukfor85** [adonisjs/core#3862] [Help] [answered]
Hello all,

I am experiencing an error when registering services in my application - require(...).default is not a constructor

I am registering services for IOC like so:

```typescript
this.app.container.singleton('Internal/ReportingService', () => {
    return new (require('../app/Services/ReportingService').default)();
});
```

There are ambient modules for TS and services are imported the following way:
```typescript
import ReportingService from '@ioc:Internal/ReportingService';
```

Services look like that:
```typescript
export default class ReportingService {
   public async someMethod() {}
}
```

I have around 30 services that represent application business logic. Services are using Lucid models.

Some of the services bindings are OK but some throw the error. I would appreciate it if someone could point me to where the issue might be.

**Comment by Barbapapazes** [comment]
You need to use import instead of require no? (there is no default in require)

**Reply by lukfor85** [reply]
AFAIK no... I can't use import inside register method because it's not async. Top level imports don't work by design, I tried it but I am getting different errors... I suspect that issue is with the fact that I am using services in other services and vice versa.

EDIT: top level imports generate the following error
```
 E_IOC_LOOKUP_FAILED: Cannot resolve "Adonis/Lucid/Orm" namespace from the IoC Container
```

**Comment by RomainLanz** [comment, marked as answer]
Hey @lukfor85 ! 👋🏻 

I am not sure what you are trying to do.
You should use providers only to register third-party packages or part of your application you want to externalize (in a package). The thing you bind should have no relation with your application code (or anything loaded by the IOC).

Here, you are binding internal services for no reason.

If your goal is to create a singleton of your service, you can simply export an instance of your class inside your module.

```ts
class MySingleton {}

const instance = new MySingleton()
export default instance
```

**Reply by thetutlage** [reply]
Exactly, as @RomainLanz said. There is no need to add an in-direction layer within your application codebase.

Providers are meant for packages distributed via npm.

**Reply by lukfor85** [reply]
Thank you, guys, for a quick response. What I want to achieve, is to have my services mockable/fakeable for testing. That's why I am trying to use AdonisJS IOC for internal services. I somehow managed to get it into a working state. The trouble was/is that some services call each other and that's the probable reason for the issue. That might be a warning for me that I have cyclic dependencies which is not a good pattern. Now it works OK. 

You are saying, that IOC is not meant for internal services. I am rewriting the existing codebase to achieve fully testable code. Maybe I need to use a different approach for that, but currently, I don't know how to achieve my goal by a different way.
